hist/hist/src/TFormula.cxx

```diff
@@ -3347,12 +3348,15 @@ Double_t TFormula::DoEval(const double * x, const double * params) const
    if (!fClingInitialized && fLazyInitialization) {
       // try recompiling the formula. We need to lock because this is not anymore thread safe
       R__LOCKGUARD(gROOTMutex);
-      auto thisFormula = const_cast<TFormula*>(this);
-      thisFormula->ReInitializeEvalMethod();
-   }
-   if (!fClingInitialized) {
-      Error("DoEval", "Formula has error and  it is not properly initialized ");
-      return TMath::QuietNaN();
+      // check again in case another thread has initialized the formula (see ROOT-10994)
```

Comment:
Should fClingInitialized be atomic?

Reply:
I think fClingInitialized should be  protected by the gROOTMutex above. 
It would be great if you can try of this fixes the problem

Reply:
It looks like the update of `fClingInitialized` is protected by the mutex by not the reads (and they probably should not).  
If this is the case, then indeed `fClingInitialized` should be an atomic.

Also with this pattern (atomic read and behind-a-lock write), one need might need to check after entering the critical section whether another thread updated it in the mean time. (because 2 threads may have concurrently check the value and found it false and both try to initialize, the lock will (as wanted) serialize the initialize but the check-after-taking the lock should be done unless doing the initialization twice is both harmless and cheap.